# Module 72.1- Forward Contract Valuation

<hr>

In [43]:
# Import libraries
import math
# import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt
# import seaborn as sns
# from sklearn.model_selection import train_test_split
# from pandas.core.common import random_state
# from sklearn.linear_model import LinearRegression

## Define Reciprocal
def reciprocal(n):
    return 1.0 / n

### LOS 72.a: Explain how the value and price of a forward contract are determined at initiation, during the life of the contract, and at expiration.

Consider a forward contract that is initially priced at its no-arbitrage value of $F_0(T) = S_0(1 + Rf)^T$. At initiation, the value of such a forward is:
\
\
$V_0(T) = S_0 – F_0(T) (1 + Rf)^{–T} = 0$
\
\
At any time during its life, the value of the forward contract to the buyer will be $V_t(T) = S_t– F_0(T) (1 + Rf)^{–(T–t)}$. This is simply the current spot price of the asset minus the present value of the forward contract price.

This value can be realized by selling the asset short at St and investing $F_0(T) (1 + Rf)^{–(T–t)}$ in a pure discount bond at Rf. These transactions end any exposure to the forward; at settlement, the proceeds of the bond will cover the cost of the asset at the forward price, and the asset can be delivered to cover the short position.

At expiration, time $T$, 
\
\
$\text{the value of a forward to the buyer is} = S_T – F_0(T)(1 + Rf)^{–(T–T)} = S_T – F_0(T)$. 
\
\
The long buys an asset valued at $S_T$ for the forward contract price of $F_0(T)$, 
\
\
gaining if $S_T > F_0(T)$, losing if $S_T < F_0(T)$. 
\
\
If the forward buyer has a gain, the forward seller has an equal loss, and vice versa.
\
\
In the more general case, when there are costs and benefits of holding the underlying asset, the value of a forward to the buyer at time $= t < T$ is:
\
\
$V_t(T) = [S_t+ PV_t \text{(costs)} – PV_t(\text{benefit)}] – F_0(T)(1 + Rf)^{–(T–t)}$

### LOS 72.b: Explain how forward rates are determined for interest rate forward contracts and describe the uses of these forward rates.

Forward rates are yields for future periods. The rate of interest on a 1-year loan to be made two years from today is a forward rate.

The notation for forward rates must identify both the length of the loan period and how far in the future the money will be loaned (or borrowed). 1y1y or $F_{1,1}$ is the rate for a 1-year loan one year from now; 2y1y or $F_{2,1}$ is the rate for a 1-year loan to be made two years from now; the 2-year forward rate three years from today is 3y2y or $F_{3,2}$; and so on.


<IMG SRC ="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/Fig521_forward_rates.jpg?raw=true"></IMG >

For money market rates the notation is similar, with 3m6m denoting a 6-month rate three months in the future.

Recall that spot rates are zero-coupon rates. We will denote the YTM (with annual compounding) on a zero-coupon bond maturing in $n$ years as $Z_n$.

An **implied forward rate** is the forward rate for which the following two strategies have the same yield over the total period:

* Investing from $t = 0$ to the forward date, and rolling over the proceeds for the period of the forward.
* Investing from $t = 0$ until the end of the forward period.

#### As an example, 
\
lending for two years at $Z_2$ would produce the same ending value as lending for one year at $Z_1$ and, at $t = 1$, lending the proceeds of that loan for one year at $F_{1,1}$. That is, $(1 + Z_2)2 = (1 + Z_1)(1+ F_{1,1})$. When this condition holds, $F_{1,1}$ is the implied (no-arbitrage) forward rate.
\
\
*implied forward rate*
\
<IMG SRC ="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/Fig522_implied_forward_rates.jpg?raw=true"></IMG>

#### Example: Implied forward rate

Consider two zero-coupon bonds, one that matures in two years and one that matures in three years, when $Z_2 = 2\%$ and $Z_3 = 3\%$. Calculate the implied 1-year forward rate two years from now, $F_{2,1}$.

In [177]:
# Implied forward rate example

Z2 = 100 * 1.02**2
print(round(Z2,2))

Z3 = 100 * 1.03**3
print(round(Z3,2))

F21 = ((109.27 / 104.04) - 1) * 100
print(round(F21,2))

104.04
109.27
5.03


#### Answer:
As illustrated in  **Implied Forward Rate**, lending for three years at $Z_3$ should be equivalent to lending for two years at $Z_2$ and then for the third year at $F_{2,1}$.


Lending $\$100$ for two years at $Z_2$ (2\%) results in a payment of $\$100x(1.02)^2= \$104.04$ 
\
at $t = 2$, 
\
\
while lending $\$100$ for three years at $Z_3$ (3\%) results in a payment of $ \$100 (1.03)^3 = \$109.27$. 
\
\
The forward interest rate $F_{2,1}$ must be $\large\frac{109.27}{104.04} \small – 1 = 5.03\%$, 
\
\
the implied forward rate from t = 2 to t = 3.


An example of an interest rate derivative is a **forward rate agreement (FRA)**, in which the fixed-rate payer (long) will pay the forward rate on a notional amount of principal at a future date, and the floating-rate payer will pay a future reference rate times that same amount of principal. In practice, only the net amount is exchanged.

Consider a 3-month forward on a 6-month MRR $\large(F_{3m6m})$ with a notional principal of 
\
\
1 million dollars. At settlement in three months, the buyer receives (or pays) the present value of
\
\
$$
\frac{\text{(realized 6-month MRR – 1% )}}{2} x \text{ 1 million dollars}
$$

We divide by 2 because MRRs are typically annualized rates. We take the present value of the difference in interest because the settlement payment is at the beginning of the 6-month period, whereas the interest savings would be at the end of the period.


<hr>

#### Example: 3-month 6month forward rate price with only 3 and 9 month MRR
Assume that the current 3-month MRR is 1.0\% and 9-month MRR is 1.2\%. Adjusting for periodicity, the no-arbitrage condition for the value of $\large{F_{3m6m}}$ is:


$$
1 + 0.012 \Big(\frac{9}{12}\Big)= \Biggr[1+0.01 \Big(\frac{3}{12} \Big) \Biggr] \Biggr[1+ \large{F_{3m,6m}}\small \Big(\frac{6}{12}\Big)\Biggr]
$$
\
\
The implied forward rate $\large{F_{3m6m}}$ as an annualized rate, is:
\
\
$$
\large{F_{3m6m}} = \Biggr[\frac{1+0.012(\frac{9}{12})}{1+0.01 (\frac{3}{12})} -1 \Biggr] x \frac{12}{6} = 0.013
$$




In [107]:
# Assume that the current 3-month MRR is 1.0% and 
# the 9-month MRR is 1.2%. Adjusting for periodicity,
# the no-arbitrage condition for the value of F3m6m is:

# annualize3m = 0.0025 + 1
annualize3m = 1.0025

# annualize9m = 0.0009 + 1
annualize9m = 1.009

# 6-month forward rate in percent 
f6m = ((annualize9m/annualize3m) -1) *100 #make % 
print("Fixed rate or price of 6 month FRA at T6 = %", round(f6m,6))

# Now we need to annualize that 3 months to 6 months
annualize6m = f6m * 2
F3m6m = round(annualize6m,2)
print("Fixed rate or price of our the FRA at T0 = %", round(annualize6m,5))
print("Fixed rate or price of our the FRA at T0 = %", F3m6m)

Fixed rate or price of 6 month FRA at T6 = % 0.648379
Fixed rate or price of our the FRA at T0 = % 1.29676
Fixed rate or price of our the FRA at T0 = % 1.3


<hr>

#### Example: Fixed-rate payoff
Now let's examine the payoff to the fixed-rate payer in an $\large{F_{3m6m}}$ FRA with a notional principal of 1 million dollars when the 6-month MRR three months from now is 1.5\%. Because the realized 6-month MRR is greater than the forward rate, the fixed-rate payer (floating-rate receiver) will have a gain.


The payment to the fixed-rate payer is the present value (discounted at 6-month MRR) of the interest differential between two 6-month loans, one at 1.3\% and one at 1.5\% (both annualized rates). The fixed-rate payer in the FRA receives:

$$
\text{1 million dollars  } x \Biggr(\frac{0.015-0.013}{2} \Biggr) \Biggr(\frac{1}{1+ \Big(0.0015 \big{/}2\Big)} \Biggr)= \$992.56
$$

FRAs are used primarily by financial institutions to manage the volatility of their interest-sensitive assets and liabilities. FRAs are also the building blocks of interest rate swaps over multiple periods. An FRA is equivalent to a single-period swap. Multiple-period swaps are used primarily by investors and issuers to manage interest rate risk.

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/2.png?raw=true">


<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/7.png?raw=true">

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/4.png?raw=true">

In [156]:
# EXAMPLE:  FRA PAYOFFS
# Given F3,6 = 1.3% and notional principal of $1mm
# If 6 month MRR 3 months from now is 1.5%

# Notional ammount of 1-million dollars
NA = 1000000

# At T9 we want to know what FRA is
# recieve the floating and pay the fixed

# FRA Floating-rate payer pays at month 9 (T9):
FRA_Float_PayerT9 = (0.015 - 0.013) * (6/12) * NA
print("The payoff end of month 9 float payer FRA: T9 = $", FRA_Float_PayerT9)

# FRA Fixed-rate payer pays at month 9 (T9):
FRA_Fixed_PayerT9 = (0.013-0.010) * (6/12) * NA 
print("The payoff end of month 9 fixed payer FRA: T9 = $", round(FRA_Fixed_PayerT9,2))
print("")
print("That is profit from the long FRA party and loss for short FRA")

print("")
print("short party doesn't pay long party at T9")
print("instead it's going to be discounted back for 9months")
print("It will be settled by the 2-counterparties at the")
print("Settled at the expiry of the FRA at T3 we settle")
print("We will discount at the MRR which is 1.5% at T3")
print("we are going to discount at the annualized rate of 1.0075%")
print("From T9 to T3 we have the annualized discount rate 1.0075%\n")

# FRA discounted to T3 rate
FRA_T3_rate = 1 + (.001 * 6/12)

FRA_T3 = 1000 / FRA_T3_rate
print("The amount payed by short party to long party = $", round(FRA_T3,2))
print("Long party is paying fixed and recieving floating\n")
print("What you're paying - what you're recieving\n") 
print("floating minus fixed annualized and that gives us")
print("that is also the amount payed by the Short FRA payer")


print("\n")
print("What if the MRR = 1.0% instead of 1.5%?")

print("\n")
print("Rate at FRA F3m6m in  %",F3m6m)
payoff_F3m6m = (0.01-0.013) * (6/12) * NA
print("\n")
print("Payoff of FRA F3m6m at T9 $",round(payoff_F3m6m,2))
print("\n")

payoff_F6 = payoff_F3m6m / 1.005
print("Discount payoff from T9 to FRA Expiry\n where the long party has to pay the short party $", round(payoff_F6,2))
print("In other words: Fixed-rate payer pays floating-rate payer.")

The payoff end of month 9 float payer FRA: T9 = $ 1000.0
The payoff end of month 9 fixed payer FRA: T9 = $ 1500.0

That is profit from the long FRA party and loss for short FRA

short party doesn't pay long party at T9
instead it's going to be discounted back for 9months
It will be settled by the 2-counterparties at the
Settled at the expiry of the FRA at T3 we settle
We will discount at the MRR which is 1.5% at T3
we are going to discount at the annualized rate of 1.0075%
From T9 to T3 we have the annualized discount rate 1.0075%

The amount payed by short party to long party = $ 999.5
Long party is paying fixed and recieving floating

What you're paying - what you're recieving

floating minus fixed annualized and that gives us
that is also the amount payed by the Short FRA payer


What if the MRR = 1.0% instead of 1.5%?


Rate at FRA F3m6m in  % 1.3


Payoff of FRA F3m6m at T9 $ -1500.0


Discount payoff from T9 to FRA Expiry
 where the long party has to pay the short party $ -1492.5

<img src ="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/3.png?raw=true">

<img src ="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/9.png?raw=true">

<img src="https://github.com/PachaTech/CFA-Level-1/blob/main/Derivatives/Module%2072/pics/12.png?raw=true">

<hr>

#### Example: Forward contract value

Long position in one-year forward contract with a price 35 dollars, 
RF= 0.03.  After 9 months, the spot price of the underlying asset = 36 dollars.  What is the value of the contract?

Where:
$F_0(T) = 35$\
$S_t = 36$\
$T = 9 \text{  months}$\
\
$V_t(T) = S_t-F_0(T) / (1 + R_f)^{T-t}$
\
\
$ = 36 - 35 / (1.03)^{(1-0.075)}= 1.26$

In [164]:
recipricol1 = (1.03**(0.25))
print(recipricol1)

# Convert to a discount factor by recipricol
recipricol2 = reciprocal(recipricol1)
print(recipricol2)

print(35 * recipricol2)
print(35 * (recipricol2*-1))

# Our answer is 35 * negative Reciprico1 + 36
answer = 35 * (recipricol2*-1)+36
print("\n")
print("Vt(T) = ", round(answer,2))

1.0074170717777329
0.9926375361451396
34.742313765079885
-34.742313765079885


Vt(T) =  1.26


<img src="">

<hr>

#### Long FRA  = pay fixed, recieve floating
at Settlement
* If MRR > fixed: Long recieves `[MRR - fixed] x Notional amt`
* If MRR < fixed: Long pays `[fixed - MRR] x Notional amt`

#### Short FRA = pay floating, recieve fixed
at Settlement
* If MRR > fixed: short pays `[MRR - fixed] x Notional amt`
* If MRR < fixed: short recieves `[fixed - MRR] x Notional amt`

<hr>

#### FRA Uses (Example)

* A corporation that expects to **borrow** 1mm for 6-months in three months time can fix borrowing cost with <u>pay-fixed</u> position in a FRA with $F_{3,6}=1.3\%$   -- here you're antipating the rates to up down in three months time so you take a **Long FRA** position 
    * Pay fixed and recieve floating.


* A corporation that expects to **lend** 1mm for 6 months in three months time can fix lending rate with <u>pay-floating</u> position in FRA with $F_{3,6}=1.3\%$   -- here you're antipating the rates to go down in three months time so you take a **Short FRA** position. 
    * Pay floating and Recieve fixed.


<hr>


# Reading 72 QBank & Quiz


<hr>

**Question 1**\
Two parties agree to a forward contract to exchange 100 shares of a stock one year from now for $\$72$ per share. Immediately after they initiate the contract, the price of the underlying stock increases to $\$74$ per share. This share price increase represents a gain for:



**Answer**\
the buyer

**Explanation**\
If the value of the underlying is greater than the forward price, this increases the value of the forward contract, which represents a gain for the buyer and a loss for the seller. 
\
(Module 72.1, LOS 72.a)

<hr>

**Question 2**\
The forward rate 2y3y represents the interest rate on a loan for the period from:

**Answer**\
Year 2 to Year 5.

**Explanation**\
2y3y is the 2-year forward 3-year rate, covering a period that begins two years from now and extends for three years after that. 
\
(Module 72.1, LOS 72.b)


<hr>

**Question 3**\
Given zero-coupon bond yields for 1, 2, and 3 years, an analyst can least likely derive an implied:


**Answer**\
2-year forward 2-year rate.

**Explanation**\
The forward rate $\large F_{2,2}$ extends four years into the future and cannot be derived using zero-coupon yields that only extend three years. From zero-coupon bond yields for 1, 2, and 3 years, we can derive implied forward rates $\large F_{1,1}$, $\large F_{1,2}$, and $\large F_{2,1} $. 

\
(Module 72.1, LOS 72.b)

<hr>

**Question 4**\
For an underlying asset that has no holding costs or benefits, the value of a forward contract to the long during the life of the contract is the:


**Answer**\
spot price minus the present value of the forward price

**Explanation**\
During the life of a forward contract on an underlying asset with no holding costs or benefits, the value to the long equals the spot price minus the present value of the forward price:
\
\
$NCC = 0$
\
$V_t(T) = S_t – F_0(T) / (1 + R_f)^{–(T–t)}$.
\
\
or
\
\
$V_t(T) = S_t - \Large\frac{F_0(T)}{1+R_f}^{(T-t)}$
\
\
(Module 72.1, LOS 72.b)

<hr>


**Question 5**\
At time $t$, prior to its settlement date at time $T$, the value $V_t$ of a long forward with a price of $F$ will be related to the spot price, $S$, of an asset that has a zero net cost of carry by:

**Answer**\
$V_t(T) = S_t – F_0(T) / (1 + R_f)^{–(T–t)}$.

**Explanation**\
The value of a long position in a forward contract prior to settlement (expiration) is:
\
$\qquad\large V_t = S_t − F_0(T)\times(1 + Rf)^{–(T – t)}$
\
\
\
when the net cost of carry is zero.\
\
(Module 72.1, LOS 72.b)

<hr>

**Question 6**\
The value of a forward or futures contract is:

**Answer**\
equal to the spot price at expiration.

**Explanation**\
The value of a forward or futures contract is typically zero at initiation, and at expiration is the difference between the spot price and the contract price. The price of a forward or futures contract is defined as the price specified in the contract at which the two parties agree to trade the underlying asset on a future date.
\
(Module 72.1, LOS 72.a)

<hr>

**Question 7**\
The most likely use of a forward rate agreement is to:

**Answer**\
lock in an interest rate for future borrowing or lending.

**Explanation**\
The purpose of a forward rate agreement (FRA) is to manage interest rate risk by locking in an interest rate for future borrowing or lending. An FRA is a forward commitment rather than a contingent claim. An interest rate swap is used to exchange a floating-rate obligation for a fixed-rate obligation.
\
(Module 72.1, LOS 72.b)

<hr>